File: A3/Sem5/LFTC/ciordales/Labs/Laborator 5/src/main/java/LexicalAtomTable.java
import java.util.HashMap;
import java.util.Map;

public class LexicalAtomTable {

    class ItemLexicalAtom{
        private String lexicalAtom;
        private Integer atomCode;

        public ItemLexicalAtom(String lexicalAtom, Integer atomCode) {
            this.lexicalAtom = lexicalAtom;
            this.atomCode = atomCode;
        }
    }

    private Map<String, Integer> lexicalTable;
    private Integer increment = 1;

    public LexicalAtomTable(){
        lexicalTable = new HashMap<String, Integer>();
        lexicalTable.put("ID", 0);
        lexicalTable.put("CONST", 1);
    }

    public void addLexicalAtom(String lexicalAtom){
        if(lexicalTable.get(lexicalAtom) == null) {
            increment++;
            lexicalTable.put(lexicalAtom, increment);
        }
    }

    public Integer getCodeAtom(String lexicalCode){
        return lexicalTable.get(lexicalCode);
    }

    public Integer getIdentifierAtomCode(){
        return 0;
    }

    public Integer getLiteralAtomCode(){
        return 1;
    }

    @Override
    public String toString() {
        String text = "";
        Map<Integer, String> hashMap = new HashMap<Integer, String>();
        for(Map.Entry<String, Integer> entry: lexicalTable.entrySet()){
            hashMap.put(entry.getValue(), entry.getKey());
        }

        for (int i = 0; i < lexicalTable.size(); i++) {
            String line = " " + hashMap.get(i) + " " + i + "\n";
            text = text + line;
        }
        return text;
    }
}

File: A3/Sem5/LFTC/lab5Clion/lexicalAtomTable.cpp
#include "lexicalAtomTable.h"


void LexicalAtomTable::addLexicalAtom(const string& lexicalAtom) {

	if (lexicalTable.find(lexicalAtom) == lexicalTable.end()) {
		lexicalTable[lexicalAtom] = this->increment;
        this->increment = this->increment + 1;
	}
}

string LexicalAtomTable::toString() {
    string text = "";
    map<int, string> hashMap;
    map<string, int>::iterator it;
    for (it=lexicalTable.begin();it!=lexicalTable.end();it++) {
        hashMap[it->second] = it->first;
    }

    for (int i = 0; i < lexicalTable.size(); i++) {
        string line = " " + hashMap[i] + " " + to_string(i) + "\n";
        text = text + line;
    }
    return text;
}